docs/batch-metrics-api.md

```diff
@@ -0,0 +1,247 @@
+
+*Figure 1. Retaining more metricsdb files*
+
+The second issue is the limited granularity of the data. We can simply expand the granularity by allowing users to query the raw data for each metric rather than responding with an aggregate of that data. One drawback of this approach is that the response size will drastically increase. This prevents us from supplying a “nodes=all” parameter like with the metrics API, as it would require a single node to gather that data from all the other nodes and retain that data in memory before responding to the user. So, users will have to query individual nodes in order to gather performance metrics from those nodes.
+
+All the raw metrics data from a period of time may be too high granularity for some users. One approach to limiting this granularity might be to respond with some aggregate of that data, like with the metrics API. However, a more compute-efficient, practical, and useful approach would be to instead respond with a sample of that raw data. The available datapoints are currently in the form of 5s granularity bins (each metricsdb file collects 5s of data). We can efficiently sample from these bins by providing users with a “sampling-period” parameter. We would partition the available data according to the sampling period and simply respond with the data from the first bin in each partition.
+
+![Sampling from bins](/docs/images/batch-metrics-api-2.png)
+
+*Figure 2. Sampling from bins*
+
+## API
+
+**performance-analyzer.properties**
+* retention-period: The number of minutes worth of metrics data to collect. The configured retention period can be read via the “/_opendistro/performanceanalyzer/batch/config” api. Default=7 (7 minutes), min=1, max = 60.
+  * Note, the default is 7 minutes because a typical use-case would be to query for 5 minutes worth of data from the node. In order to do this, a client would actually select a starttime of now-6min and an endtime of now-1min (this one minute offset will give sufficient time for the metrics in the time range to be available at the node). Atop this 6 minutes of retention, we need an extra 1 minute of retention to account for the time that would have passed by the time the query arrives at the node, and for the fact that starttime and endtime will be rounded down to the nearest sampling-period.
+
+**API**
+
+Queries:
+
+* POST \<endpoint\>:9200/_opendistro/performanceanalyzer/batch/config -H ‘Content-Type: application/json’ -d ‘{“enabled”: true}’
+* POST \<endpoint\>:9200/_opendistro/performanceanalyzer/batch/cluster/config -H ‘Content-Type: application/json’ -d ‘{“enabled”: true}’
+* GET \<endpoint\>:9200/_opendistro/_performanceanalyzer/_agent/batch?metrics=\<metrics\>&starttime=\<starttime\>&endtime=\<endtime\>&samplingperiod=\<samplingperiod\>
+* GET \<endpoint\>:9600/_opendistro/_performanceanalyzer/batch?metrics=\<metrics\>&starttime=\<starttime\>&endtime=\<endtime\>&samplingperiod=\<samplingperiod\>
+
```

Comment:
We should add a section about enabling and what it will cost and disabling (how to). Will they require restarts etc.

Reply:
See above